ipq807x: add Arcadyan AW1000 support

aiamadeus opened the pull request:
Hardware specification:
  SoC: Qualcomm IPQ8072A
  Flash: Toshiba NAND 1GiB
  RAM: 1 GiB of DDR3 466 MHz
  Ethernet: 4x 1Gbps + 1x 2.5Gbps
  WiFi1: QCN5024 2.4GHz ax 4x4
  WiFi2: QCN5054 5GHz ax 4x4
  Button: WiFi, WPS, Reset
  Modem: RG500Q-EA
  USB: 1 x USB 3.0
  Power: DC 12V 4A

Flash instructions:
  1. Download the initramfs image, rename it to initramfs.bin, and host it with the tftp server.
  2. Interrupt U-Boot and run these commands:
     tftpboot initramfs.bin
     bootm
  3. After openwrt boots up, use scp or luci web to upload sysupgrade.bin to upgrade.

aiamadeus committed:
ipq807x: add Arcadyan AW1000 support

Hardware specification:
  SoC: Qualcomm IPQ8072A
  Flash: Toshiba NAND 1GiB
  RAM: 1 GiB of DDR3 466 MHz
  Ethernet: 4x 1Gbps + 1x 2.5Gbps
  WiFi1: QCN5024 2.4GHz ax 4x4
  WiFi2: QCN5054 5GHz ax 4x4
  Button: WiFi, WPS, Reset
  Modem: RG500Q-EA
  USB: 1 x USB 3.0
  Power: DC 12V 4A

Flash instructions:
  1. Download the initramfs image, rename it to
     initramfs.bin, and host it with tftp server.
  2. Interrupt U-Boot and run these commands:
     tftpboot initramfs.bin
     bootm
  3. After openwrt boots up, use scp or luci web
     to upload sysupgrade.bin to upgrade.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>
Reviewed-by: Robert Marko <robimarko@gmail.com>

aiamadeus commented:
```
[    0.000000] Kernel command line: console=ttyMSM0,115200n8 pcie_dev=RG50 root=/dev/ubiblock0_1
[    0.000000] Unknown kernel command line parameters "pcie_dev=RG50", will be passed to user space.
[   11.200228] ath11k c000000.wifi: qmi ignore invalid mem req type 3
[   11.207683] ath11k c000000.wifi: chip_id 0x0 chip_family 0x0 board_id 0x270 soc_id 0xffffffff
[   11.207721] ath11k c000000.wifi: fw_version 0x290104a5 fw_build_timestamp 2023-08-02 20:32 fw_build_id WLAN.HK.2.9.0.1-01890-QCAHKSWPL_SILICONZ-1
[   11.238548] ath11k c000000.wifi: failed to fetch board data for bus=ahb,qmi-chip-id=0,qmi-board-id=624,variant=Arcadyan-AW1000 from ath11k/IPQ8074/hw2.0/board-2.bin
[   11.238611] ath11k c000000.wifi: failed to fetch board data for bus=ahb,qmi-chip-id=0,qmi-board-id=624 from ath11k/IPQ8074/hw2.0/board-2.bin
[   11.252368] ath11k c000000.wifi: failed to fetch board data for bus=ahb,qmi-chip-id=0,qmi-board-id=624 from ath11k/IPQ8074/hw2.0/board-2.bin
[   11.264941] ath11k c000000.wifi: failed to fetch board.bin from IPQ8074/hw2.0
[   11.277517] ath11k c000000.wifi: qmi failed to fetch board file: -12
[   11.284522] ath11k c000000.wifi: failed to load board data file: -12
```
I don't know why the qmi-board-id value changed to 624. BDF shows:
```
FileSize: 131180
FileCRC32: 094fc8bf
FileMD5: 2c4f4405a14802e8762e127740841c45
BoardNames[0]: 'bus=ahb,qmi-chip-id=0,qmi-board-id=255,variant=Arcadyan-AW1000'
BoardLength[0]: 131072
BoardCRC32[0]: 0c14fee9
BoardMD5[0]: 50ac399d338e51667fe5076aad94a4c5
```

robimarko commented:
This could be the first device where they actually fused the `board-id` like everyone is supposed to do.
So you need to adjust the `board-id` value for the board-2.bin as 255 is the default in case vendor did not fuse the ID.

robimarko commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -28,6 +28,7 @@ endef
 # <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
 
 ALLWIFIBOARDS:= \
+	arcadyan_aw1000 \
```
Did you make a PR for this BDF?

Cause this wont work as its not present in the BDF repo.

aiamadeus replied:
I will, I'm just wondering about qmi-board-id changed itself.

robimarko replied:
Well, it cannot really change itself as the FW will try and read the fuse on boot.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-aw1000.dts:
```diff
@@ -0,0 +1,348 @@
+		};
+	};
+
+	led_spi {
```
Please, no underscores in node names.

Same applies for all nodes.

robimarko commented on target/linux/qualcommax/ipq807x/base-files/lib/upgrade/platform.sh:
```diff
@@ -51,9 +60,6 @@ platform_do_upgrade() {
 		nand_do_restore_config || nand_do_upgrade_failed
 		buffalo_upgrade_optvol
 		;;
-	dynalink,dl-wrx36)
```
I get what you are trying to do here by cleaning all of these that use the same, but please separate commit.

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -0,0 +1,18 @@
+
+case "$board" in
+	arcadyan,aw1000)
+		addr=$(cat /sys/class/net/wan/address)
```
Are the MAC-s populated in ART?

aiamadeus replied:
No, there is no mac address from the wireless zone on ART.

robimarko replied:
Hm, I wonder if we should add support for patching the MAC-s in like ath10k has during calibration loading instead of later fixing them up

robimarko commented on package/kernel/qca-ssdk/patches/0003-qca8081-add-LED-rule-for-Arcadyan-AW1000.patch:
```diff
@@ -0,0 +1,90 @@
+-#define QCA808X_PHY_MMD7_LED0_CTRL_ENABLE       0x8670
+-#define QCA808X_PHY_MMD7_LED1_CTRL_DISABLE      0x0
+-#define QCA808X_PHY_MMD7_LED2_CTRL_DISABLE      0x0
++#define QCA808X_PHY_MMD7_LED_POLARITY_ACTIVE_LOW  0x6
```
Why are these being renamed?

aiamadeus replied:
It looks like the CTRL_ENABLE and CTRL_DISABLE values are the same for LED 1/2/3.

robimarko replied:
This patch wont apply anymore as QCA changed the API so they now have MMD modify directly

aiamadeus replied:
Update, although it seems a bit redundant, is it possible now?

robimarko replied:
We are going to drop SSDK QCA8081 driver anyway as @Ansuel has added LED support to the upstream driver.

So, it would just add code that is going to get dropped rather soon.

aiamadeus commented:
```
[   10.947954] ath11k c000000.wifi: ipq8074 hw2.0
[   10.951437] ath11k c000000.wifi: FW memory mode: 0
[   10.956131] remoteproc remoteproc0: powering up cd00000.q6v5_wcss
[   10.960618] remoteproc remoteproc0: Booting fw image IPQ8074/q6_fw.mdt, size 668
[   11.312294] remoteproc remoteproc0: remote processor cd00000.q6v5_wcss is now up
[   11.356300] ath11k c000000.wifi: qmi ignore invalid mem req type 3
[   11.368504] ath11k c000000.wifi: chip_id 0x0 chip_family 0x0 board_id 0x270 soc_id 0xffffffff
[   11.375922] ath11k c000000.wifi: fw_version 0x290104a5 fw_build_timestamp 2023-08-02 20:32 fw_build_id WLAN.HK.2.9.0.1-01890-QCAHKSWPL_SILICONZ-1
[   11.456687] qcom-q6v5-wcss-pil cd00000.q6v5_wcss: fatal error received: 
[   11.456687] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.9.0.1-01890-QCAHKSWPL_SILICONZ-1
[   11.456687] Image Variant : IMAGE_VARIANT_STRING=8074.wlanfw.eval_v2Q
[   11.456687] 
[   11.456687] phyrf_bdf.c:629 Assertion 0 failedparam0 :zero, param1 :zero, param2 :zero.
[   11.456687] Thread ID      : 0x00000072  Thread name    : wlan_platform  Process ID     : 0
[   11.456687] Register:
[   11.456687] SP : 0x4b5d3090
[   11.456687] FP : 0x4b5d3098
[   11.456687] PC : 0x4b1080c4
[   11.456687] SSR : 0x00000008
[   11.456687] BADVA : 0x00020000
[   11.456687] LR : 0x4b107860
[   11.456687] 
[   11.456687] Stack Dump
[   11.456687] from : 0x4b5d3090
[   11.456687] to   : 0x4b5d35b0
[   11.456687] 
[   11.503586] remoteproc remoteproc0: crash detected in cd00000.q6v5_wcss: type fatal error
[   11.525810] remoteproc remoteproc0: handling crash #1 in cd00000.q6v5_wcss
[   11.534185] remoteproc remoteproc0: recovering cd00000.q6v5_wcss
[   11.566744] remoteproc remoteproc0: stopped remote processor cd00000.q6v5_wcss
```
It seems that this may happen if the wireless part of  ART does not have a valid mac address.

aiamadeus commented:
Compare with other ipq807x which ART does not have mac address:
ART on SPI: 11_fix_wifi_mac works
ART on NAND: 11_fix_wifi_mac not work

aiamadeus committed:
kernel: add support for Toshiba TH58NYG3S0HBAI4

Correct oob size from 128 to 256 for Toshiba TH58NYG3S0HBAI4 flash.
Since it is not ONFI compliant NAND, the model name cannot be read
from anywhere, add a static NAND ID entry to correct this.

However, the NAND ID of this flash is inconsistent with the datasheet.
The actual NAND ID is only 4 ID bytes, the last ID byte is missing.[1]

Maybe this flash is counterfeit, or maybe it's another problem.
Another Toshiba flash had the same problem before. Refer to commit
a83dc6b ("kernel: move Toshiba-TC58NVG0S3H patch to ipq40xx"), put
the patch into qualcommax target to avoid affecting other devices.

The patch is verified on Arcadyan AW1000.

[1] Datasheet available at (the ID table is on page 50):
https://europe.kioxia.com/content/dam/kioxia/newidr/productinfo/datasheet/201910/DST_TH58NYG3S0HBAI4-TDE_EN_31565.pdf

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>
Reviewed-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
The NAND fix should go upstream, would be ideal to link the datasheet in the commit as well.

aiamadeus commented:
Thanks, I have sent the NAND fix patch to upstream kernel. BDF pr also created.

hauke commented:
Please rebase these changes on top of current master.

aiamadeus commented:
I actually tried and failed. From the comparison of the art partition, the
checksum of ath11k is located at 0xa. But I don't know how to calculate
it, starting from ipq807x there are two mac address. A failed example:
```
ath11k_patch_mac1() {
	local mac=$1
	local target=$2

	caldata_patch_mac "$mac" 0xe 0xa "$target"
}

ath11k_patch_mac2() {
	local mac=$1
	local target=$2

	caldata_patch_mac "$mac" 0x14 "" "$target"
}
```

aiamadeus committed:
qualcommax: cleanup whitespace and sort-order

Replace blanks with tabs, also sort base-files alphabetically.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>
Reviewed-by: Robert Marko <robimarko@gmail.com>

aiamadeus commented:
@Ansuel Sorry but can you please take a look and merge this?

robimarko commented:
Can you rebase on top of the master and apply the same logic as in this commit:
https://github.com/openwrt/openwrt/commit/947b44d9ae17b4d56578c10a5dd675e79e4c621b

aiamadeus commented:
> Can you paste here the full bootlog without the NAND patch?

https://pastebin.com/EgJcY3XZ
> This is the patch I have so far, can you please test it as well:

Thank you so much. Should I keep changes to nand_ids.c?
If we don't keep it, the driver can't find TH58NYG3S0HBAI4.
If we keep it this works.

robimarko commented:
Yeah, I forgot this is not ONFI compliant NAND so the model name is not read from anywhere and needs an entry in the ID table, so the ONFI fixup hook is pretty much useless.

This should work, and enables the optimized timings:
```
From 72901bc50248d7b7cc4a5ea868691c81863c9ad1 Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Sat, 11 Nov 2023 10:26:20 +0100
Subject: [PATCH] mtd: rawnand: kioxia: Add support for TH58NYG3S0HBAI4

TH58NYG3S0HBAI4 is 8Gbit 8bit parallel NAND, but its not ONFI compliant.
Default OOB calculation will calculate the OOB as 128 bytes, but according
to the datasheet[1] OOB size is 256 bytes.

Timings in the datasheet[1] match those of TH58NVG2S3HBAI4, so just reuse
those.

[1] https://europe.kioxia.com/content/dam/kioxia/newidr/productinfo/datasheet/201910/DST_TH58NYG3S0HBAI4-TDE_EN_31565.pdf

Signed-off-by: Robert Marko <robimarko@gmail.com>
---
 drivers/mtd/nand/raw/nand_ids.c     | 3 +++
 drivers/mtd/nand/raw/nand_toshiba.c | 4 +++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/mtd/nand/raw/nand_ids.c b/drivers/mtd/nand/raw/nand_ids.c
index 650351c62af6..0c60248da4c4 100644
--- a/drivers/mtd/nand/raw/nand_ids.c
+++ b/drivers/mtd/nand/raw/nand_ids.c
@@ -64,6 +64,9 @@ struct nand_flash_dev nand_flash_ids[] = {
 	{"TH58NVG3S0HBAI4 8G 3.3V 8-bit",
 		{ .id = {0x98, 0xd3, 0x91, 0x26, 0x76} },
 		  SZ_4K, SZ_1K, SZ_256K, 0, 5, 256, NAND_ECC_INFO(8, SZ_512)},
+	{"TH58NYG3S0HBAI4 8G 1.8V 8-bit",
+		{ .id = {0x98, 0xa3, 0x91, 0x26, 0x76} },
+		  SZ_4K, SZ_1K, SZ_256K, 0, 5, 256, NAND_ECC_INFO(8, SZ_512)},
 
 	LEGACY_ID_NAND("NAND 4MiB 5V 8-bit",   0x6B, 4, SZ_8K, SP_OPTIONS),
 	LEGACY_ID_NAND("NAND 4MiB 3,3V 8-bit", 0xE3, 4, SZ_8K, SP_OPTIONS),
diff --git a/drivers/mtd/nand/raw/nand_toshiba.c b/drivers/mtd/nand/raw/nand_toshiba.c
index d3d34d71921f..f4aede3caae2 100644
--- a/drivers/mtd/nand/raw/nand_toshiba.c
+++ b/drivers/mtd/nand/raw/nand_toshiba.c
@@ -290,7 +290,9 @@ static int toshiba_nand_init(struct nand_chip *chip)
 	if ((!strncmp("TH58NVG2S3HBAI4", chip->parameters.model,
 		     sizeof("TH58NVG2S3HBAI4") - 1)) ||
 	    (!strncmp("TH58NVG3S0HBAI4", chip->parameters.model,
-		     sizeof("TH58NVG3S0HBAI4") - 1)))
+		     sizeof("TH58NVG3S0HBAI4") - 1)) ||
+	    (!strncmp("TH58NYG3S0HBAI4", chip->parameters.model,
+		     sizeof("TH58NYG3S0HBAI4") - 1)))
 		th58nvg2s3hbai4_init(chip);
 
 	return 0;
-- 
2.41.0
```

aiamadeus commented:
> This should work, and enables the optimized timings:

Thanks, but this won't work. As I submitted before:
```
+	{"TH58NYG3S0HBAI4 8G 1.8V 8-bit",
+		{ .id = {0x98, 0xa3, 0x91, 0x26, 0x76} },
+		  SZ_4K, SZ_1K, SZ_256K, 0, 4, 256, NAND_ECC_INFO(8, SZ_512) },
```
Otherwise its oob size is still 128.

aiamadeus commented:
> I am worried that we maybe another case of:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/drivers/mtd/nand/raw/nand_ids.c?h=next-20231114&id=2c5947cffd81ac8181346efacdca3c777ab330ba

Yes, you are right. The last ID is missing.
```
[    0.842987] nand: NAND ID len: 4
[    0.843768] nand: ID[0]: 0x98
[    0.847204] nand: ID[1]: 0xa3
[    0.850068] nand: ID[2]: 0x91
[    0.853018] nand: ID[3]: 0x26
```

robimarko commented:
Well, that is a shame.

That leaves only the option of adding the entry with 4 ID bytes and leaving a good comment above it, as well as detailed commit description as to why is it required.

robimarko commented:
Can you further expand on the reason why this NAND needs a special ID entry, because its not really clear as to why in the commit message, especially emphasize that its probably counterfeit, and link the datasheet of original part.

aiamadeus commented:
> Can you further expand on the reason why this NAND needs a special ID entry, because its not really clear as to why in the commit message, especially emphasize that its probably counterfeit, and link the datasheet of original part.

Sorry, it should look better now?

robimarko commented:
Yeah, looks better now.

I will do another pass tommorow.

robimarko commented:
LGTM,
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

robimarko commented:
@Ansuel Can you merge this?

There are others ready as well, I am doing another round of reviews

neheb commented:
@aiamadeus Does this seem correct?

```
--- a/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-aw1000.dts
+++ b/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-aw1000.dts
@@ -46,23 +46,11 @@
                        gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
                };
 
-               lte-power {
-                       gpio-export,name = "lte_power";
-                       gpio-export,output = <1>;
-                       gpios = <&tlmm 30 GPIO_ACTIVE_HIGH>;
-               };
-
                lte-reset {
                        gpio-export,name = "lte_reset";
                        gpio-export,output = <1>;
                        gpios = <&tlmm 63 GPIO_ACTIVE_HIGH>;
                };
-
-               usb-vbus {
-                       gpio-export,name = "usb_vbus";
-                       gpio-export,output = <0>;
-                       gpios = <&tlmm 9 GPIO_ACTIVE_LOW>;
-               };
        };
 
        keys {
@@ -160,6 +148,23 @@
                        gpios = <&led_gpio 11 GPIO_ACTIVE_HIGH>;
                };
        };
+
+       lte_vbus: regulator-usb0-vbus {
+               compatible = "regulator-fixed";
+               regulator-name = "lte_vbus";
+               regulator-min-microvolt = <5000000>;
+               regulator-max-microvolt = <5000000>;
+               gpios = <&tlmm 30 GPIO_ACTIVE_HIGH>;
+               enable-active-high;
+       };
+
+       usb_vbus: regulator-usb0-vbus {
+               compatible = "regulator-fixed";
+               regulator-name = "usb_vbus";
+               regulator-min-microvolt = <5000000>;
+               regulator-max-microvolt = <5000000>;
+               gpios = <&tlmm 9 GPIO_ACTIVE_LOW>;
+       };
 };
 
 &tlmm {
@@ -217,10 +222,14 @@
 
 &qusb_phy_0 {
        status = "okay";
+
+       vdda-phy-supply = <&lte_vbus>;
 };
 
 &qusb_phy_1 {
        status = "okay";
+
+       vdda-phy-supply = <&usb_vbus>;
 };
 
 &ssphy_0 {

```

aiamadeus commented:
No, the lte-power gpio controls the lte module, not the usb. Also it should be `vdd-supply = <&xxx>`.

neheb commented:
I assume that LTE module is USB based.

Is it correct to set the USB power GPIO to the second phy?

aiamadeus commented:
It works like this:
```
&qusb_phy_1 {
	vdd-supply = <&usb_vbus>;
	status = "okay";
};
```

aiamadeus commented:
It works like this:
```
&qusb_phy_1 {
	vdd-supply = <&usb_vbus>;
	status = "okay";
};
```

neheb commented:
It's interesting that rt-ax89x.dts uses GPIO 30 for USB and it's used for LTE here.
